backend/ce/api/stats.py

```diff
@@ -21,18 +21,105 @@
 model_id2:
     ...}
 '''
+
+import os.path
+
+from modelmeta import DataFile
+
+import numpy as np
+import numpy.ma as ma
+from shapely.geometry import Polygon, Point, CAP_STYLE
+from shapely.wkt import loads
+from netCDF4 import Dataset
+from sqlalchemy.orm.exc import NoResultFound
+
 def stats(sesh, id_, time, area, variable):
     '''
     '''
+    try:
+        fname, = sesh.query(DataFile.filename).filter(DataFile.unique_id == id_).one()
+    except NoResultFound:
+        return {}
+
+    assert os.path.exists(fname)
+
+    nc = Dataset(fname)
+
+    polygon = loads(area)
+
+    # Mask out data that isn't inside the input polygon
+    mask = polygonToMask(nc, polygon)
+
+    assert variable in nc.variables
+
+    data = nc.variables[variable]
+
+    if time:
+        assert 'time' in nc.variables[variable].dimensions
+        data = data[time,:,:] # FIXME: Assumes 3d data... doesn't support levels
+
+    # Extend the mask into the time dimension (if it exists)
+    mask = np.repeat(mask, data.size / mask.size).reshape(data.shape)
+
+    data = ma.masked_array(data, mask=mask)
+
     return {
-        'model_id1':
+        id_:
         {
-            'min': 0.0,
-            'max': 2.0,
-            'mean': 1.0,
-            'median': 1.0,
-            'stdev': 1.0,
-            'units': 'degC'
+            'min': np.min(data),
+            'max': np.max(data),
+            'mean': np.mean(data),
+            'median': ma.median(data)[0],
+            'stdev': np.std(data),
+            'units': nc.variables[variable].units,
+            'ncells': data.compressed().size
         },
     }
 
+
+def pointInPoly(x, y, poly):
+    if x is ma.masked or y is ma.masked:
+        return ma.masked
+
+    cell = Point(x, y).buffer(2, cap_style=CAP_STYLE.square)
```

Comment:
Why is a buffer of size 2 applied? Doesn't this just create a circle of radius 2 degrees? If we want to evaluate intersection based on an expanded cell centroid, the buffer should be some ratio of distance between centroids.